### PR TITLE
MetaInterval: Prevent restarting an interval if it has already finished and we don't specify a start time

### PR DESCRIPTION
**Pull request by JaceMayer**
## Issue description
As per issue #1203, when a finished sequence has ```resume()``` called, ```Func()``` have their callbacks called. This PR aims to mitigate this unintuitive design decision.

## Solution description
For MetaIntervals, if you call ```.resume()``` and the interval is already in the finished state, it will return out of the function, unless you have specified a ```startT```.  For consistency, this check has been placed in MetaInterval as opposed to overriding the function for just ```Sequence()```

## Checklist
I have done my best to ensure that…
* [x] …I have familiarized myself with the CONTRIBUTING.md file
* [x] …this change follows the coding style and design patterns of the codebase
* [x] …I own the intellectual property rights to this code
* [x] …the intent of this change is clearly explained
* [ ] …existing uses of the Panda3D API are not broken
* [ ] …the changed code is adequately covered by the test suite, where possible.

### CODE REVIEW

**Review by rdb — CHANGES_REQUESTED** [review]
I think this is a fine solution, thanks!  But it is a change of behaviour, so it really needs a docstring explaining the behaviour (and explaining how it changes as of 1.11.0, which will be the next release that a behaviour-changing change will land in).

I'm wondering if we should also put a similar check in the base Interval class for consistency.  Resuming just a solitary instantaneous interval like `Func(print, 'test')` will also rerun upon calling `resume()` when it's at the end, so it should probably behave the same way as `Sequence(Func(print, 'test'))`, right?

We then also should make the change to the C++ CInterval class (ie. so that `resume()` without start_t argument doesn't do anything for a finished interval).

A unit test would be a nice-to-have.

**Comment by DavidACrompton** [comment]
> I'm wondering if we should also put a similar check in the base Interval class for consistency. Resuming just a solitary instantaneous interval like `Func(print, 'test')` will also rerun upon calling `resume()` when it's at the end, so it should probably behave the same way as `Sequence(Func(print, 'test'))`, right?

Been looking into this @rdb, the behaviour for `Func(print, 'test')`  and `Sequence(Func(print, 'test'))`, are different for some reason.
```
from direct.interval.IntervalManager import ivalMgr
from direct.interval.IntervalGlobal import *
from direct.task.TaskManagerGlobal import taskMgr

sut = Func(print, 'test')
sut.start()
ivalMgr.step()
taskMgr.step()
print(sut.getState())
```
This returns 0 (S_initial), where as changing `sut = Func(print, 'test')` to `sut = Sequence(Func(print, 'test'))` will have the print statement return 3 (S_final) as expected.

Perhaps FunctionInterval should be changed to have this - so it would better match the flow of a Sequence/Parallel?
```
    def privInstant(self):
        self.state = CInterval.SStarted
        # Evaluate the function
        self.function(*self.extraArgs, **self.kw)
        # Print debug information
        self.notify.debug(
            'updateFunc() - %s: executing Function' % self.name)
        self.state = CInterval.SFinal
        self.intervalDone()
```

File: direct/src/interval/cInterval.cxx [hunk 1]
```diff
@@ -175,9 +175,14 @@ pause() {
...
  * Restarts the interval from its current point after a previous call to
  * pause().
+ * In versions previous to 1.11, this would resume an interval even if it had ended
+ * As of 1.11 we now check if the interval has ended, and only resume it if startT has been passed
  */
 void CInterval::
 resume() {
+    if(get_state() == S_final){
+        return;
+    }
```
Comment (rdb) [inline direct/src/interval/cInterval.cxx:185]: Style nitpicks: indentation, spaces around parentheses in if condition, so:
```cpp
  if (get_state() == S_final) {
    return;
}
```

File: direct/src/interval/FunctionInterval.py [hunk 2]
```diff
@@ -86,11 +86,18 @@ def makeUniqueName(func, suffix = ''):
         return name
 
     def privInstant(self):
+        """
+        In versions previous to 1.11, this function did not set the state of the Interval
+        In 1.11 this has been changed to be consistent with MetaInterval
+        """
```
Comment (rdb) [inline direct/src/interval/FunctionInterval.py:92]: IMO for the purposes of generating the API documentation it's not great to have a docstring on a private method, I'd be OK with omitting this or turning it into a regular comment

File: direct/src/interval/Interval.py [hunk 3]
```diff
@@ -134,6 +134,12 @@ def pause(self):
         return self.getT()
 
     def resume(self, startT = None):
+        """
+        In versions previous to 1.11, this would resume an interval even if it had ended
+        As of 1.11 we now check if the interval has ended, and only resume it if startT has been passed
+        """
```
Comment (rdb) [inline direct/src/interval/Interval.py:140]: PEP8 nitpicks:
> For flowing long blocks of text with fewer structural restrictions (docstrings or comments), the line length should be limited to 72 characters.

> Block comments generally consist of one or more paragraphs built out of complete sentences, with each sentence ending in a period.

**Review by rdb — CHANGES_REQUESTED** [review]
I think I agree with the FunctionInterval change for consistency with the base implementation.

Thanks!  I just have some nitpicks.

**Pull request closed without merging**